### PR TITLE
Documentation for OQL expressions

### PR DESCRIPTION
**Pull request by pijuskri**
- Documentation for OQL expressions following the new structure
- Includes aggregates and parameters
- Needs team review
- To be merged with other OQL documentation MR's

### CODE REVIEW

File: content/en/docs/refguide/modeling/data/oql/oql-expressions.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+Aggregations are functions that reduce a list of values from a retrieved column (or columns) into a singular value. They can be used as an attribute in a `SELECT` clause or in a condition in `HAVING`. 
+
+When combined with a `GROUP BY` clause, aggregations can be used with any column normally available in `SELECT`, not just those specified in the `GROUP BY` clause. Aggregation is performed over groups separately in this case.
+
+## Syntax
+
+```sql
+  COUNT ( * )
+  | { COUNT | AVG | MAX | MIN | SUM } ( [ DISTINCT ] attribute_path )
```
Comment (dahfjkg) [inline content/en/docs/refguide/modeling/data/oql/oql-expressions.md]: Let's add `STRING_AGG` here

File: content/en/docs/refguide/modeling/data/oql/oql-expressions.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+```sql
+SELECT Sum(Stock) AS ProductCount FROM Sales.Product
+```
+| ProductCount |
+|:------------:|
+|     103      |
+
+The sum of all products in stock:
+```sql
+SELECT SUM(Stock) AS StockSum FROM Sales.Product
```
Comment (dahfjkg) [inline content/en/docs/refguide/modeling/data/oql/oql-expressions.md]: This example is repeated twice

File: content/en/docs/refguide/modeling/data/oql/oql-expressions.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+The highest stock number:
+```sql
+SELECT MAX(Stock) as StockMax FROM Sales.Product
+```
+| StockMax |
+|:--------:|
+|    54    |
+
+Using `MAX` on a string column gives the last alphabetically sorted name:
```
Comment (dahfjkg) [inline content/en/docs/refguide/modeling/data/oql/oql-expressions.md]: TBH, this is probably too technical

File: content/en/docs/refguide/modeling/data/oql/oql-expressions.md [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+```sql
+SELECT Name
+FROM Module.Person
+WHERE
+    Age > $param 
+    AND
+    Active = true
+```
+
+If the value of `$param` is not provided, the query will be equivalent to:
```
Comment (dahfjkg) [inline content/en/docs/refguide/modeling/data/oql/oql-expressions.md]: I think this is not entirely correct. If a parameter is empty, that part of the expression should be replaced with true, like this:
```
WHERE
    true 
    AND
    Active = true
```
This is even more relevant in case of `OR`, so maybe we should use `OR` in our examples

**Review by dahfjkg — CHANGES_REQUESTED** [review]

File: content/en/docs/refguide/modeling/data/oql/oql-expressions.md (hunk 3 above)
Comment (pijuskri) [reply content/en/docs/refguide/modeling/data/oql/oql-expressions.md]: Agreed, i dont think the example is neccesary for more advanced users and this behaviour is mentioned in the documentation

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md [hunk 5]
```diff
@@ -0,0 +1,264 @@
+---
+title: "OQL Expressions"
+url: /refguide/oql-expressions-new/
+---
+
+## Introduction
+
+An OQL expression is a query building block that returns a value or a list of values. Expressions can be one of the following:
+
+* a constant
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:10]: Can this also be a list of constants? Or can only some expressions return lists?

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+* a function
+* a combination of attribute names, constants, and functions connected by operators
+* a subquery
+
+OQL expressions can be used in `WHERE`, `SELECT`, `GROUP BY`, `UNION`, and `JOIN` clauses.
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md]: Should probably link to the OQL Clauses doc here.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+
+## Aggregations
+
+Aggregations are functions that reduce a list of values from a retrieved column (or columns) into a single value. They can be used in the following ways:
+
+* as an attribute in a `SELECT` clause
+* as a condition in a `HAVING` clause
+* When combined with a `GROUP BY` clause, aggregations can be used with any column normally available in `SELECT`, not just those specified in the `GROUP BY` clause. Aggregation is performed over groups separately in this case.
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md]: I don't understand this at all - and there is no example. How can you put an aggregation into a GROUP BY clause if the column isn't mentioned in the GROUP BY clause?

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+
+```sql
+SELECT COUNT(*) AS ProductCount FROM Sales.Product
+```
+
+| ProductCount |
+|:------------:|
+|      4       |
+
+The same result can be retrieved by using `COUNT` on a single table:
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md]: On a single **attribute**?

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+```
+
+| Name     | Stock |
+|----------|-------|
+| Cheese   | 5     |
+| Milk     | 54    |
+| Tomatoes | 44    |
+| Tomatoes | NULL  |
+
+#### COUNT
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:96]: Should these examples be in the same order as we explain the expressions (i.e. with AVG first)?

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+
+The average stock per product:
+
+```sql
+SELECT AVG(Stock) AS ProductCount FROM Sales.Product
+```
+
+| ProductCount |
+|:------------:|
+|    34.333    |
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md]: This is only right because one of the **Tomatoes** rows has NULL. If it had a number in then the average would be wrong 😀 because we aren't filtering on distinct product names.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+
+```sql
+SELECT STRING_AGG(Name, ',') as ProductNames FROM Sales.Product
+```
+
+|         ProductNames          |
+|:-----------------------------:|
+| Cheese,Milk,Tomatoes,Tomatoes |
+
+#### Distinct
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md]: I think we should link to the clauses doc here.
Also, DISTINCT isn't an expression - the example is about COUNT so I think we should move this to the COUNT examples.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+
+| ProductCount |
+|:------------:|
+|      3       |
+
+## Parameters
+
+Parameters are external variables that are referenced to by name in an OQL query. To use a defined parameter in a query, prepend the `$` sign to the name of the parameter. 
+
+If you use undefined in `IN` and `LIKE` comparison expressions, the condition always returns `true`. In other cases, undefined parameters cause an exception.
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:219]: We don't have any documentation for `IN` and `LIKE`. Are these expressions going to be added here, or will they be in a different document?

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+```sql
+SELECT Name
+FROM Sales.Person
+WHERE
+    Age > $param 
+    OR
+    Job = 'Sales'
+```
+
+If the value of `$param` is not provided, the query will be equivalent to:
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md]: Not sure exactly what "not provided" means in this context.
Does it mean that the parameter has not been created and therefore is not in the current scope, or can you create a variable but still not initialize it.
If it is the first, then perhaps say "If the parameter `$param` has not been created"?

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+SELECT COUNT(DISTINCT Name) AS ProductCount FROM Sales.Product
+```
+
+| ProductCount |
+|:------------:|
+|      3       |
+
+## Parameters
+
+Parameters are external variables that are referenced to by name in an OQL query. To use a defined parameter in a query, prepend the `$` sign to the name of the parameter.
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:217]: Do we have some documentation on how these are created - I guess they are just Java variables, but how does that work in data sets?
Perhaps this would become obvious if I was actually creating some OQL?

**Review by MarkvanMents — COMMENTED** [review]
Hi @pijuskri and @dahfjkg 
I've proofread the whole of this document.
I think the style matches the clauses document quite well, I'm not worried that they will seem too different.
I've asked some questions for clarification. 
I'll go back to the clauses document while you digest my changes and come back to me.
Let me know if you want to set up a meeting.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md (hunk 13 above)
Comment (pijuskri) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md]: The query can reference a parameter and then the parameter mapping needs to be provided when calling the query, say in a java action. I wouldn't say "created" is the best word to use here, parameters are quite temporary.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md (hunk 12 above)
Comment (pijuskri) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:219]: Yes they will be documented in `oql-expression-syntax`

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md (hunk 9 above)
Comment (pijuskri) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:96]: Yes, did not notice the order, will change it.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md (hunk 7 above)
Comment (pijuskri) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md]: I misformated this. It should not be part of the "ways to use aggregations" list. 
The comment about group by is for aggregates the `SELECT` clause. Ill clarify this

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md (hunk 5 above)
Comment (pijuskri) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:10]: There can be a list of of constants, but only when used with the `IN` operator. Don't think that single use case is important enough to mention here. Some expressions (like subquery) can return multiple values, so its useful to mention that here.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md (hunk 14 above)
Comment (pijuskri) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:217]: The best we have now is the datasets documentation. There arent any visuals or examples there and honestly I dont know how that works either.
For java actions its partially explained in the [API docs](https://apidocs.rnd.mendix.com/8/runtime/com/mendix/systemwideinterfaces/connectionbus/requests/types/ITextGetRequest.html#createParameterMap()).

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md (hunk 5 above)
Comment (MarkvanMents) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:10]: OK, Thanks.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md (hunk 12 above)
Comment (MarkvanMents) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:219]: OK, thanks.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md (hunk 13 above)
Comment (MarkvanMents) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md]: So could we say
```suggestion
If the value of `$param` is not provided as a parameter to the query, the query will be equivalent to:
```

**Review by MarkvanMents — COMMENTED** [review]
I've responded to some of your replies.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md (hunk 10 above)
Comment (pijuskri) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md]: Thats true. Making this query distinct would require a subquery, so a change of the description is better.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md (hunk 13 above)
Comment (pijuskri) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md]: Yes that seems more clear

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+An OQL expression is a query building block that returns a value or a list of values. Expressions can be one of the following:
+
+* a constant
+* a function
+* a combination of attribute names, constants, and functions connected by operators
+* a subquery
+
+OQL expressions can be used in `WHERE`, `SELECT`, `GROUP BY`, `UNION`, and `JOIN` clauses. For more information, see [OQL clauses](/refguide/oql-clauses/).
+
+## Aggregations
```
Comment (dahfjkg) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md]: I'm referring to this in Clauses using tag #aggregates. Please add it here

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+---
+
+## Introduction
+
+An OQL expression is a query building block that returns a value or a list of values. Expressions can be one of the following:
+
+* a constant
+* a function
+* a system variable
+* a combination of attribute names, constants, system variables and functions connected by operators
```
Comment (dahfjkg) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md]: A subquery can also be a part of "a combination of attribute names, constants, ...", so I'd put this line last and mention subquery in it

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+* a subquery
+
+OQL expressions can be used in `WHERE`, `SELECT`, `GROUP BY`, `UNION`, and `JOIN` clauses. For more information, see [OQL clauses](/refguide/oql-clauses/).
```
Comment (dahfjkg) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md]: 1. I think `HAVING` too;
2. For `JOIN` clauses, I'd rephrase it as "and `ON` conditions of `JOIN` clauses"

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md [hunk 18]
```diff
@@ -0,0 +1,265 @@
...
+
+```sql
+SELECT COUNT(Stock) AS ProductCount FROM Sales.Product
+```
+
+| ProductCount |
+|:------------:|
+|      3       |
+
+There are duplicate values in the `Name` column, which might not want to be counted separately. [Distinct](/refguide/oql-clauses/#Distinct) can be used to get the number of unique rows:
```
Comment (dahfjkg) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md]: Let's explain how `DISTINCT` works with `AVG` and `SUM` too

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md [hunk 19]
```diff
@@ -0,0 +1,265 @@
...
+* as an attribute in a `SELECT` clause
+* as a condition in a `HAVING` clause 
+
+When combined with a `GROUP BY` clause, aggregations in `SELECT` can be used with any column normally available, not just those specified in the `GROUP BY` clause. Aggregation is performed over groups separately in this case.
+
+### Syntax
+
+```sql
+  COUNT ( * )
+  | { COUNT | AVG | MAX | MIN | SUM | STRING_AGG } ( [ DISTINCT ] attribute_path )
```
Comment (dahfjkg) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md]: This syntax is not entirely correct: `DISTINCT` is allowed with all functions except `STRING_AGG`

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md [hunk 20]
```diff
@@ -100,7 +103,17 @@ SELECT AVG(Stock) AS ProductCount FROM Sales.Product
 
 | ProductCount |
 |:------------:|
-|    34.333    |
+|    36.75     |
+
+There are duplicate values in the `Stock` column, which can be ignored by using [DISTINCT](/refguide/oql-clauses/#Distinct). The query below returns the average unique stock:
+
+```sql
+SELECT AVG(Distinct Stock) AS ProductCount FROM Sales.Product
```
Comment (dahfjkg) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md]: Let's use more representative aliases for different aggregate functions. Both in queries and in result tables.

**Comment by dahfjkg** [comment]
Passed technical review

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md (hunk 9 above)
Comment (MarkvanMents) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:96]: I've reordered them as the initial description has it ordered COUNT (*) and then COUNT | AVG | MAX - so it seems better to present the descriptions in this order.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md (hunk 14 above)
Comment (MarkvanMents) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:217]: Thanks for the explanation. I've added a link to the API Docs.

**Comment by dahfjkg** [comment]
This can be closed as it is merged into https://github.com/mendix/docs/pull/8431

**Comment by MarkvanMents** [comment]
No unresolved comments. Merged into #8431 so closing this PR.
